### PR TITLE
Integrity protection for live updates to WriteBatch

### PR DESCRIPTION
**Pull request by ajkr**
This PR adds the foundation classes for key-value integrity protection and the first use case: protecting live updates from the source buffers added to `WriteBatch` through the destination buffer in `MemTable`. The width of the protection info is not yet configurable -- only eight bytes per key is supported. This PR allows users to enable protection by constructing `WriteBatch` with `protection_bytes_per_key == 8`. It does not yet expose a way for users to get integrity protection via other write APIs (e.g., `Put()`, `Merge()`, `Delete()`, etc.).

The foundation classes (`ProtectionInfo.*`) embed the coverage info in their type, and provide `Protect.*()` and `Strip.*()` functions to navigate between types with different coverage. For making bytes per key configurable (for powers of two up to eight) in the future, these classes are templated on the unsigned integer type used to store the protection info. That integer contains the XOR'd result of hashes with independent seeds for all covered fields. For integer fields, the hash is computed on the raw unadjusted bytes, so the result is endian-dependent. The most significant bytes are truncated when the hash value (8 bytes) is wider than the protection integer.

When `WriteBatch` is constructed with `protection_bytes_per_key == 8`, we hold a `ProtectionInfoKVOTC` (i.e., one that covers key, value, optype aka `ValueType`, timestamp, and CF ID) for each entry added to the batch. The protection info is generated from the original buffers passed by the user, as well as the original metadata generated internally. When writing to memtable, each entry is transformed to a `ProtectionInfoKVOTS` (i.e., dropping coverage of CF ID and adding coverage of sequence number), since at that point we know the sequence number, and have already selected a memtable corresponding to a particular CF. This protection info is verified once the entry is encoded in the `MemTable` buffer.

Test Plan:
- an integration test to verify a wide variety of single-byte changes to the encoded `MemTable` buffer are caught
- add to stress/crash test to verify it works in variety of configs/operations without intentional corruption
- [deferred] unit tests for `ProtectionInfo.*` classes for edge cases like KV swap, `SliceParts` and `Slice` APIs are interchangeable, etc.

### CODE REVIEW

File: db/memtable.cc [hunk 1]
```diff
@@ -517,6 +520,58 @@ Status MemTable::Add(SequenceNumber s, ValueType type,
...
+      verify_status = Status::Corruption("Value length too short");
+    }
+    if (verify_status.ok() && verify_value_len > input.size()) {
+      verify_status = Status::Corruption("Value length too long");
+    }
+    if (verify_status.ok()) {
+      Slice verify_value = Slice(input.data(), verify_value_len);
+      memtable_kv_prot_info.SetValueChecksum(GetSliceNPHash64(verify_value));
+      verify_status = kv_prot_info->VerifyAgainst(memtable_kv_prot_info);
+    }
```
Comment (mrambacher) [inline db/memtable.cc:569]: These could be simplified slightly by making them else statements for line 555:
if (! GetVarInt )
} else if length is too short or long {
} else {
}
You could still leave the "VerifyAgainst" at line 568 outside this loop.

File: db/memtable.cc [hunk 2]
```diff
@@ -517,6 +520,58 @@ Status MemTable::Add(SequenceNumber s, ValueType type,
...
+    if (verify_status.ok() && verify_value_len < input.size()) {
+      verify_status = Status::Corruption("Value length too short");
+    }
+    if (verify_status.ok() && verify_value_len > input.size()) {
+      verify_status = Status::Corruption("Value length too long");
+    }
+    if (verify_status.ok()) {
+      Slice verify_value = Slice(input.data(), verify_value_len);
+      memtable_kv_prot_info.SetValueChecksum(GetSliceNPHash64(verify_value));
+      verify_status = kv_prot_info->VerifyAgainst(memtable_kv_prot_info);
```
Comment (mrambacher) [inline db/memtable.cc:568]: Does it matter if you validate kv_prot_info against memtable_kv_prot_info or the other way around?  Which one do you care about the flags for?

File: db/kv_checksum.cc [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+      GetSequenceNumber() != expected.GetSequenceNumber()) {
+    return Status::Corruption("sequence number mismatch");
+  }
+  if (expected.HasColumnFamilyId() &&
+      GetColumnFamilyId() != expected.GetColumnFamilyId()) {
+    return Status::Corruption("column family ID mismatch");
+  }
+  if (expected.HasValueType() && GetValueType() != expected.GetValueType()) {
+    return Status::Corruption("value type mismatch");
+  }
```
Comment (mrambacher) [inline db/kv_checksum.cc:59]: I believe this code will assert if the current flags do not contain all of the expected flags.  Was that the intent?  Or should that be a Corruption or InvalidArgument or something else?

**Comment by riversand963** [comment]
Will the per-write-batch protection info ever be persisted to WAL? If not, will memtable rebuilding during db recovery re-compute the protection info?

File: db/kv_checksum.cc (hunk 3 above)
Comment (ajkr) [reply db/kv_checksum.cc:59]: Yes, it is currently considered a programming error if we expect a field to be a particular value while that field is not present (but the other way around is a simple successful verification). One use case of this "superset" comparison is verification before we replace a field (like `ValueType` which is used differently in `WriteBatch` vs. `MemTable`; or checksum of value, which can be changed in merges after `max_successive_merges` is reached, or if an in-place update callback is invoked). Another use case is we have extra fields, like column family ID, which got added for `WriteBatch`, but aren't finally stored in `MemTable`.

A different design could be to provide APIs to replace/strip protection info for each field, while requiring the user to provide the expected old protection info so we can verify there was no corruption before replacing/removing it. If we use those correctly, the `KvProtectionInfo` comparison function can assume protection coverage is exactly the same.

File: db/memtable.cc (hunk 2 above)
Comment (ajkr) [reply db/memtable.cc:568]: `kv_prot_info` may have more info because it accumulated some extra things like CF ID along its journey here. So it should be the `this` in `VerifyAgainst()`, as that verifies `this`'s protection info is a superset of `memtable_kv_prot_info`. (See above comment for a possible design change which may make things simpler.)

**Review by ajkr — COMMENTED** [review]
> Will the per-write-batch protection info ever be persisted to WAL? If not, will memtable rebuilding during db recovery re-compute the protection info?

For now the plan is to recompute. The main concern in persisting to WAL was the need to introduce a bunch of new optypes. Also the WAL entries are protected by WAL physical fragment checksums already, so adding it there would be double protection. The downside though is it looks a bit complicated to protect the steps from the WAL fragment verification until the KV is reconstructed. I think we reconstruct `WriteBatch`es not KVs, so we may be better off using a separate checksum to protect the `WriteBatch` reconstruction, then regenerate KV checksums from that trusted `WriteBatch` buffer.

File: db/kv_checksum.cc [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  return Status::OK();
+}
+
+void KvProtectionInfo::Clear() {
+  coverage_flags_ = static_cast<CoverageFlags>(0);
```
Comment (mrambacher) [inline db/kv_checksum.cc:64]: There is a potential issue here if you are running with asserts off.  In that case, the values are not initialized, meaning they are random.  Or if you called Clear on an object that had previously been used, the values will match the old settings.  So if you call GetXXX (e.g, GetChecksum), it will return a random or old value if asserts are off.

Seems like you need to set the values to some default (or illegal value) here.

File: include/rocksdb/write_batch.h [hunk 5]
```diff
@@ -181,6 +184,13 @@ class WriteBatch : public WriteBatchBase {
     // we won't break existing clients of Handler on a source code level when
     // adding a new member function.
 
+    // Default implementation will just call `PutCF()` without
+    // `KvProtectionInfo` for backwards compatibility.
+    virtual Status PutCF(uint32_t column_family_id, const Slice& key,
+                         const Slice& value, const KvProtectionInfo*) {
```
Comment (siying) [inline include/rocksdb/write_batch.h:190]: I'm confused. This is a public interface and contains a class that is not defined in public. Do we expect users to use it?

File: db/kv_checksum.h [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+#pragma once
+
+#include "db/dbformat.h"
+#include "rocksdb/types.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+// Holds checksums of user keys/values and copies of internal metadata for
+// protecting integrity of data as it is processed.
+class KvProtectionInfo {
```
Comment (siying) [inline db/kv_checksum.h:15]: This class is very complicated. Is it possible to start with something simpler. E.g., we hard code how the checksum is generated and generated to a single checksum, rather than several checksums and each for a field?

File: db/kv_checksum.h [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+#pragma once
+
+#include "db/dbformat.h"
+#include "rocksdb/types.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+// Holds checksums of user keys/values and copies of internal metadata for
+// protecting integrity of data as it is processed.
```
Comment (siying) [inline db/kv_checksum.h:14]: Can you explain somewhere why we need those metadata along the operation? It's a lot of information to keep in the class per KV.

File: db/write_batch.cc [hunk 8]
```diff
@@ -574,64 +605,165 @@ Status WriteBatchInternal::Iterate(const WriteBatch* wb,
...
+          KvProtectionInfo kv_prot_info = wb->kv_prot_infos_[found];
+          // `ValueType` is used differently in `WriteBatch` and
+          // `WriteBatch::Handler`. Here in `WriteBatch`, different `ValueType`s
+          // are used to distinguish whether an entry specifies a column family.
+          // Later in `WriteBatch::Handler`, that distinction is dropped and a
+          // single `ValueType` is used for both cases. So, this is our last
+          // chance to verify against the distinguished `ValueType`.
+          KvProtectionInfo expected_kv_prot_info;
+          expected_kv_prot_info.SetValueType(static_cast<ValueType>(tag));
+          s = kv_prot_info.VerifyAgainst(expected_kv_prot_info);
```
Comment (siying) [inline db/write_batch.cc:622]: It appears that we anyway would verify in MemTable::Add(). Do we need this extra one?

**Review by siying — COMMENTED** [review]
It's interesting to learn that we've made progress on the public interface of KV checksum!

File: db/kv_checksum.h (hunk 7 above)
Comment (ajkr) [reply db/kv_checksum.h:14]: It's needed to verify the metadata wasn't corrupted. To save memory, we can fold all the checksums and metadata copies using XOR into a configurable number of bytes. As mentioned in the description, this PR only addresses `WriteBatch`, and does not implement such folding yet under the assumption that per-KV overhead is not important yet. Please let me know if you think that assumption is flawed (it is just my random guess).

File: db/kv_checksum.h (hunk 6 above)
Comment (ajkr) [reply db/kv_checksum.h:15]: Yes, one option was to have a checksum that covers all the key/value/timestamp buffers and all the metadata. On the design doc, Peter pointed out that would catch cases like key and value being swapped. However, it is a bit difficult to work with as the components that need to be covered change during processing. For example, we add seqno during the `WriteBatch` -> `MemTable` processing. I think we also lose CF ID at some point.

Also we'd need to add back the xxh3 streaming APIs in order to implement that efficiently. This problem is fixable though.

File: db/write_batch.cc (hunk 8 above)
Comment (ajkr) [reply db/write_batch.cc:622]: Good question, it is a pretty subtle/questionable case this could catch. Imagine value type got flipped from `kTypeColumnFamilyValue` to `kTypeValue` or vice-versa. The check in `MemTable::Add()` wouldn't catch it because `MemTableInserter::PutCF()` uses `kTypeValue` in either case.

I think such a flip wouldn't actually matter since our checks for the other components would catch it (it would likely cause CF ID to be wrong, and if not, key would definitely be wrong due to start reading it from the wrong offset). Should we remove checks for flips that don't matter? I'm not sure.

File: include/rocksdb/write_batch.h (hunk 5 above)
Comment (ajkr) [reply include/rocksdb/write_batch.h:190]: Hm, good point. We do not expect users to use it. However we need it in `MemTableInserter` which is used via a `Handler*`. I will think more how to achieve this without exposing the class name publicly... LMK any suggestion.

File: db/kv_checksum.cc (hunk 3 above)
Comment (pdillinger) [reply db/kv_checksum.cc:59]: > so we can verify there was no corruption before replacing/removing it.

I'm not necessarily saying anything you don't know, but...

Verifying data often and with precise granularity is the enemy of performance, but not necessarily the enemy of data integrity. It is definitely the friend of debugability, but not necessarily appropriate for release builds.

For example, suppose in release builds instead of storing a hash for each of n fields, we just xor hashes (independent seeds! [0]) from all the n fields that have been set. Assuming your program logic can always reconstruct which inputs should have gone in, we can verify with essentially the same level of protection as the same size hash for each field.

With that in mind, one thing to consider is statically typing the set of inputs that have gone into a KvProtectionInfo. Although in theory there could be 2^n types to work with, hopefully we don't need to go outside a typical order very often. So using single-letter abbreviations, we could have

ProtectionInfoK
ProtectionInfoKV
ProtectionInfoKVS
ProtectionInfoKVSC
ProtectionInfoKVSCT

And each has functions (value semantics, const functions) to add the next level, back out one level (release mode check deferred until all levels are backed out), or replace any level from an old value to a new value (release mode check deferred). For full verification in release mode, you get to backing out the K from ProtectionInfoK, which should probably return Status.

[0] You can't really have independent seeds for the same kind of CRC checksum. Proof available upon request. That's why I'm going to say "hashes".

File: util/hash.cc [hunk 9]
```diff
@@ -80,4 +80,19 @@ uint64_t Hash64(const char* data, size_t n) {
   return XXH3p_64bits(data, n);
 }
 
+uint64_t GetSlicePartsNPHash64(const SliceParts& data) {
+  // TODO(ajkr): use XXH3 streaming APIs to avoid the copy/allocation.
```
Comment (pdillinger) [inline util/hash.cc:84]: XXH3(p) is highly tuned for variable length known at the start of hashing or compile-time length, because its behavior varies substantially depending on length. Therefore its streaming support is complex and might not be very fast on small-to-medium data. It's funny here that we know the overall length up front; we just have a split buffer.

I would expect something with more uniform structure, like recent generations of Murmur hashes, to work better for split / streaming hashes of small-to-medium data size.

File: db/kv_checksum.h (hunk 6 above)
Comment (siying) [reply db/kv_checksum.h:15]: How about regenerate the checksum every time we add something? Or extending it?

Comment (ajkr) [reply db/kv_checksum.h:15]: I read Peter's comment (https://github.com/facebook/rocksdb/pull/7748#discussion_r541215097) after writing my reply here. I think that design provides a good way to embed the covered checksum components into the type system, without having to do extra checksum computation (as with regenerating) or worrying about the order the fields were added (as with extending).

File: db/kv_checksum.h [hunk 10]
```diff
@@ -0,0 +1,307 @@
+//  Copyright (c) 2020-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+//
+// This file contains classes containing fields to protect individual entries.
```
Comment (ajkr) [inline db/kv_checksum.h:6]: @pdillinger - I got things working based on your suggested class design. Please take a look and LMK if it's what you had in mind/any further suggestions.

Again, I assumed the protection bytes per entry overhead is not hugely important at this stage (`WriteBatch`). In the latest update, only eight bytes per key is supported. There is some work in this PR towards supporting smaller powers of two, but I hope to defer finishing it to later.

File: db/kv_checksum.h [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+  if (N < sizeof(T)) {
+    assert(sizeof(T) % N == 0);
+    for (size_t i = 0; i < sizeof(T); i += N) {
+      uint64_t info_val = 0;
+      memcpy(&info_val, info_bytes, N);
+      dst_val = dst_val ^ info_val;
+      info_bytes += N;
+    }
+  } else {
+    // TODO(ajkr): For calling `NPFoldXor()` multiple times with `info`s smaller
```
Comment (pdillinger) [inline db/kv_checksum.h:167]: That's a misleading anti-pattern to me. Although you can technically get away with one of the mix-ins not being hashed, your overall protection is really only as strong as your smallest hash being xored in.

You can rely on our stock hash functions to be lossless for (fixed) input sizes up to their output size, by the way. Ping me if you want a lossless hash construction for smaller sizes.

File: db/kv_checksum.h (hunk 10 above)
Comment (ajkr) [reply db/kv_checksum.h:6]: Do you mean to pass around the expected numerical value directly, or just replace `ProtectionInfo::buf_` with a numerical value? I think the `ProtectionInfo*` wrapper classes are useful for embedding the coverage info in the type.

If we do keep the `ProtectionInfo*` classes, I am not sure about abandoning `reinterpret_cast`. My first stab at this design included a class for each step of adding/removing protection info (i.e., `ProtectionInfo`, `ProtectionInfoK`, ..., `ProtectionInfoKVOTC`). Then, it was kind of convenient to avoid accessing the buffer as `kvot_.kvo_.kv_.k_.buf_`. It was and still is also useful for invoking the default copy constructor of a more-protected type (see any of the `Protect*()` implementations).

**Review by ajkr — COMMENTED** [review]
Thanks for the review! I understood the comments except needed clarification on one.

**Comment by pdillinger** [comment]
Weird I can't reply to your comment.

> Do you mean to pass around the expected numerical value directly, or just replace ProtectionInfo::buf_ with a numerical value?

The latter

> it was kind of convenient to avoid accessing the buffer as kvot_.kvo_.kv_.k_.buf_

Just as you're repeating UpdateK for each variant, I figured you could repeat a protected T* RawValue() for each variant.

Or perhaps there should be a "private" (implementation detail) hierarchy using inheritance to implement and inherit the RawValue & Update* functions, then a public "hierarchy", with ProtectionInfoBAR inheriting from ProtectionInfoBaseBAR (not ProtectionInfoBA) and adding Strip* and Protect*. You might need to make friends between the classes to avoid casts.

**Comment by ajkr** [comment]
> > Do you mean to pass around the expected numerical value directly, or just replace ProtectionInfo::buf_ with a numerical value?
> 
> The latter

Taking a step back, I am beginning to wonder if templating the protection info classes on integer type (or on integer width as in this PR) leads to too many difficulties. The difficulties aren't apparent in this PR since it doesn't yet expose any configurable protection info size. But, we plan to do that eventually. Then, the client code (`MemTable`, `WriteBatch`, blocks/caches) will need to use these templates. I assume we can't templatize all (or any?) of these clients, so inside them we have to translate from the dynamic bytes-per-key to the corresponding class template instantiation. Further, the class template will need to be hidden as it is passed around/stored. I have some difficult plan to try to achieve this and will write it down later. But overall, it is beginning to feel like making width dynamic everywhere, storing in unsigned char array, and dealing with endianness may be easier.

File: include/rocksdb/write_batch.h (hunk 5 above)
Comment (ajkr) [reply include/rocksdb/write_batch.h]: Removed.

File: db/kv_checksum.h (hunk 7 above)
Comment (ajkr) [reply db/kv_checksum.h]: Done, we are now only storing the XOR of hashes.

File: db/kv_checksum.cc (hunk 4 above)
Comment (ajkr) [reply db/kv_checksum.cc]: This code is removed.

File: db/write_batch.cc (hunk 8 above)
Comment (ajkr) [reply db/write_batch.cc]: Done, removed it by dropping the CF vs non-CF distinction (e.g., `kTypeColumnFamilyValue` vs `kTypeValue`) before adding to the protection info. We just use the non-CF variant everywhere.

**Review by ajkr — COMMENTED** [review]
@pdillinger - GitHub is likewise preventing me from adding replies to the rest of your comments, but they should all be addressed now.

> Or perhaps there should be a "private" (implementation detail) hierarchy using inheritance to implement and inherit the RawValue & Update* functions, then a public "hierarchy", with ProtectionInfoBAR inheriting from ProtectionInfoBaseBAR (not ProtectionInfoBA) and adding Strip* and Protect*. You might need to make friends between the classes to avoid casts.

It seems like it'd add a vtable pointer to each object, which would be prohibitively expensive. Although, we don't necessarily have to use the same type for the currently processed KV's protection info and huge vectors (not implemented yet) containing protection info for all KVs in memory. More on this soon...

File: util/hash.cc (hunk 9 above)
Comment (ajkr) [reply util/hash.cc:84]: It needs more investigation/benchmarking and possibly a new dependency, so I'll probably defer looking at other hash functions.

File: db/write_batch.cc [hunk 12]
```diff
@@ -167,9 +172,8 @@ class TimestampAssigner : public WriteBatch::Handler {
   }
 
   Status DeleteRangeCF(uint32_t, const Slice& begin_key,
-                       const Slice& end_key) override {
+                       const Slice& /* end_key */) override {
     AssignTimestamp(begin_key);
-    AssignTimestamp(end_key);
```
Comment (ajkr) [inline db/write_batch.cc:172]: This is sort of out of scope, but it seems like a mistake to assign two timestamps for start and end key. The end key is actually stored in the value when KVs are encoded. It became relevant because we can't support such an update to a part of the value in our `ProtectionInfo.*` classes.

File: db/dbformat.h [hunk 13]
```diff
@@ -145,9 +145,6 @@ inline void UnPackSequenceAndType(uint64_t packed, uint64_t* seq,
                                   ValueType* t) {
   *seq = packed >> 8;
   *t = static_cast<ValueType>(packed & 0xff);
-
-  assert(*seq <= kMaxSequenceNumber);
```
Comment (ajkr) [inline db/dbformat.h:149]: This change is needed for testing corrupted memtable entries would see a `Status::Corruption` (otherwise they'd sometimes hit this assertion failure).

Comment (pdillinger) [reply db/dbformat.h:149]: assertions are useful code documentation. Consider leaving them as comments, like

    // Removed for corruption work: assert(*seq <= kMaxSequenceNumber);

File: db/blob/blob_file_cache.cc [hunk 14]
```diff
@@ -23,7 +23,8 @@ BlobFileCache::BlobFileCache(Cache* cache,
                              uint32_t column_family_id,
                              HistogramImpl* blob_file_read_hist)
     : cache_(cache),
-      mutex_(kNumberOfMutexStripes, GetSliceNPHash64),
+      mutex_(kNumberOfMutexStripes,
+             static_cast<uint64_t (*)(const Slice&)>(&GetSliceNPHash64)),
```
Comment (pdillinger) [inline db/blob/blob_file_cache.cc:27]: On first inspection, I thought this was UB but it appears legit. https://en.cppreference.com/w/cpp/language/overloaded_address

Consider stuffing in a constant in hash.h, without the static_cast, perhaps `kGetSliceNPHash64UnseededFnPtr`. But OK as is.

**Review by pdillinger — APPROVED** [review]
LGTM. Copying here some updated thoughts on requirements or potential scope:

For filter queries and other correctness-critical hashing operations (some memtable and index variants?) we might want a 'verifiable' hash of the key tracked with the key. The hash can be computed as early as possible as protection info and re-verified (optionally) at each step where it is used as a hash. That's something to consider as this work evolves.

File: db/blob/blob_file_cache.cc (hunk 14 above)
Comment (ajkr) [reply db/blob/blob_file_cache.cc:27]: Hid it as suggested. I also only learned this feature after adding an overload caused a compiler error.

**Review by ajkr — COMMENTED** [review]
Thanks for the review!

> For filter queries and other correctness-critical hashing operations (some memtable and index variants?) we might want a 'verifiable' hash of the key tracked with the key. The hash can be computed as early as possible as protection info and re-verified (optionally) at each step where it is used as a hash. That's something to consider as this work evolves.

I didn't quite understand last time but got it now. By using the same hash function as filter, we can verify the filter query's hash for free, or skip the hash to save computation.
